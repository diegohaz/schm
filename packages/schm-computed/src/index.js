// @flow
/**
 * Add computed parameters to schema.
 */
const computed = (params: Object) => (previous: any) =>
  previous.merge({
    parse(...args) {
      const parsed = previous.parse(...args);

      Object.keys(params).forEach(name => {
        const fn = params[name];
        if (typeof fn !== "function") {
          throw new Error(`[schm-computed] ${name} must be a function`);
        }

        parsed[name] = fn(parsed);
      });

      return parsed;
    }
  });

module.exports = computed;
