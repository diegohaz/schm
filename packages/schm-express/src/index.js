// @flow
import schema from "schm";

const convertEmptyToTrue = (object: Object): Object =>
  Object.keys(object).reduce(
    (finalObject, key) => ({
      ...finalObject,
      [key]: object[key] === "" ? true : object[key]
    }),
    {}
  );

/**
 * Returns an express middleware that validates and parses querystring based
 * on a given schema.
 * @example
 * const express = require('express')
 * const { query } = require('schm-express')
 *
 * const app = express()
 *
 * // request /?foo&bar=1&bar=baz
 * app.get('/', query({ foo: Boolean, bar: [String] }), (req, res) => {
 *   console.log(req.query) // { foo: true, bar: ['1', 'baz'] }
 * })
 */
export const query = (params: Object) => (
  req: Object,
  res: Object,
  next: Function
) => {
  // $FlowFixMe
  const querySchema = schema(params);
  const values = convertEmptyToTrue(req.query);
  querySchema
    .validate(values)
    .then(parsed => {
      req.query = parsed;
      next();
    })
    .catch(errors => {
      req.schmError = true;
      next(errors);
    });
};

/**
 * Returns an express middleware that validates and parses request body based
 * on a given schema.
 * @example
 * const express = require('express')
 * const bodyParser = require('body-parser')
 * const { body } = require('schm-express')
 *
 * const app = express()
 * app.use(bodyParser.json())
 * app.use(bodyParser.urlencoded({ extended: false }))
 *
 * // send { foo: 1, bar: 'baz' }
 * app.post('/', body({ foo: Boolean, bar: [String] }), (req, res) => {
 *   console.log(req.body) // { foo: true, bar: ['baz'] }
 * })
 */
export const body = (params: Object) => (
  req: Object,
  res: Object,
  next: Function
) => {
  // $FlowFixMe
  const bodySchema = schema(params);
  bodySchema
    .validate(req.body)
    .then(parsed => {
      req.body = parsed;
      next();
    })
    .catch(errors => {
      req.schmError = true;
      next(errors);
    });
};

/**
 * Handles errors from schm-express.
 * @example
 * const express = require('express')
 * const { query } = require('schm-express')
 *
 * const app = express()
 *
 * // request / without querystring
 * app.post('/', query({ foo: { type: Boolean, required: true } }), (req, res) => {
 *   ...
 * })
 *
 * app.use(errorHandler())
 *
 * // it will respond with 400 and error descriptor in response body
 */
export const errorHandler = () => (
  err: any,
  req: Object,
  res: Object,
  next: Function
) => {
  if (req.schmError) {
    return res.status(400).json(err);
  }
  return next(err);
};
