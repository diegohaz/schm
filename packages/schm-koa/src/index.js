// @flow
import schema from 'schm'

const convertEmptyToTrue = (object: Object): Object =>
  Object.keys(object).reduce(
    (finalObject, key) => ({
      ...finalObject,
      [key]: object[key] === '' ? true : object[key],
    }),
    {},
  )

/**
 * Returns a koa middleware that validates and parses querystring based
 * on a given schema.
 * @example
 * const Koa = require('koa')
 * const { query } = require('schm-koa')
 *
 * const app = new Koa()
 *
 * app.use(query({ foo: Boolean, bar: [String] }))
 *
 * // request /?foo&bar=1&bar=baz
 * app.use((ctx) => {
 *   console.log(ctx.state.query) // { foo: true, bar: ['1', 'baz'] }
 * })
 */
export const query = (params: Object) => async (
  ctx: Object,
  next: Function,
) => {
  const querySchema = schema(params)
  const values = convertEmptyToTrue(ctx.query)
  try {
    ctx.state.query = await querySchema.validate(values)
    next()
  } catch (e) {
    ctx.state.schmError = true
    throw e
  }
}

/**
 * Returns a koa middleware that validates and parses request body based
 * on a given schema.
 * @example
 * const Koa = require('koa')
 * const bodyParser = require('koa-bodyparser')
 * const { body } = require('schm-koa')
 *
 * const app = new Koa()
 *
 * app.use(bodyParser())
 * app.use(body({ foo: Boolean, bar: [String] }))
 *
 * // send { foo: 1, bar: 'baz' }
 * app.use((ctx) => {
 *   console.log(ctx.state.body) // { foo: true, bar: ['baz'] }
 * })
 */
export const body = (params: Object) => async (ctx: Object, next: Function) => {
  const bodySchema = schema(params)
  try {
    ctx.state.body = await bodySchema.validate(ctx.request.body)
    next()
  } catch (e) {
    ctx.state.schmError = true
    throw e
  }
}

/**
 * Handles errors from schm-koa.
 * @example
 * const Koa = require('koa')
 * const { query, errorHandler } = require('schm-koa')
 *
 * const app = new Koa()
 *
 * app.use(errorHandler())
 * app.use(query({ foo: { type: Boolean, required: true } }))
 *
 * // request / without querystring
 * app.use((ctx) => {
 *   ...
 * })
 *
 * // it will respond with 400 and error descriptor in response body
 */
export const errorHandler = () => async (ctx: Object, next: Function) => {
  try {
    await next()
  } catch (e) {
    if (ctx.state.schmError) {
      ctx.body = e
      ctx.status = 400
    }
  }
}
