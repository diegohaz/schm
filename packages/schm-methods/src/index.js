// @flow

/**
 * Adds methods to schm parsed object.
 */
const methods = (params: Object) => (previous: any) =>
  previous.merge({
    parse(...args) {
      const parsed = previous.parse(...args)

      Object.keys(params).forEach(name => {
        const fn = params[name]
        if (typeof fn !== 'function') {
          throw new Error(`[schm-methods] ${name} must be a function`)
        }
        parsed[name] = fn.bind(parsed, parsed)
      })

      return parsed
    },
  })

module.exports = methods
