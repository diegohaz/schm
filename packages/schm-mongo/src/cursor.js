const cursor = () => {}

export default cursor
