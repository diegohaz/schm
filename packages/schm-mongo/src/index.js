export query from './query'
export fields from './fields'
export page from './page'
export near from './near'
