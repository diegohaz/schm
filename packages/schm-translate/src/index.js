// @flow
import get from 'lodash/get'

/**
 * Translates values keys to schema keys.
 */
const translate = (params: Object) => (previous: any) =>
  previous.merge({
    parse(values = {}) {
      const parsedValues = Object.keys(params).reduce(
        (finalValues, param) => ({
          ...finalValues,
          [param]: get(finalValues, params[param]),
        }),
        values,
      )

      return previous.parse(parsedValues)
    },
  })

module.exports = translate
