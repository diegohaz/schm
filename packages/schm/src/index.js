import schema, { group } from './schema'
import parse from './parse'
import validate from './validate'
import mapValues from './mapValues'

module.exports = schema
module.exports.group = group
module.exports.parse = parse
module.exports.validate = validate
module.exports.mapValues = mapValues
