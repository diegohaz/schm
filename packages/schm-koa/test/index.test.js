import Koa from 'koa'
import bodyParser from 'koa-bodyparser'
import request from 'supertest'
import { query, body, errorHandler } from '../src'

const servers = []

afterAll(() => servers.map(server => server.close()))

const createApp = middleware => {
  const app = new Koa()
  app.use(bodyParser())
  app.use(errorHandler())
  app.use(middleware)
  app.use(ctx => {
    ctx.body = ctx.state
  })
  const server = app.listen()
  servers.push(server)
  return server
}

describe('query', () => {
  it('handles ?foo as foo=true', async () => {
    const schema = query({ foo: [Boolean] })
    const app = createApp(schema)
    const response = await request(app).post('?foo')
    expect(response.body.query.foo).toEqual([true])
  })

  it('handles array in querystring', async () => {
    const schema = query({ foo: [String] })
    const app = createApp(schema)
    const response = await request(app).post('?foo=1&foo=baz')
    expect(response.body.query.foo).toEqual(['1', 'baz'])
  })

  it('handles errors', async () => {
    const schema = query({ foo: { type: Boolean, required: true } })
    const app = createApp(schema)
    const response = await request(app).post('/')
    expect(response.status).toBe(400)
    expect(response.body).toMatchSnapshot()
  })
})

describe('body', () => {
  it('handles schema', async () => {
    const schema = body({ foo: Boolean })
    const app = createApp(schema)
    const response = await request(app)
      .post('/')
      .send({ foo: 1 })
    expect(response.body.body.foo).toBe(true)
  })

  it('handles errors', async () => {
    const schema = body({ foo: { type: Boolean, required: true } })
    const app = createApp(schema)
    const response = await request(app).post('/')
    expect(response.status).toBe(400)
    expect(response.body).toMatchSnapshot()
  })
})

describe('errorHandler', () => {
  it('passes through', async () => {
    const app = createApp(() => Promise.reject())
    const response = await request(app).get('/')
    expect(response.status).toBe(404)
  })
})
