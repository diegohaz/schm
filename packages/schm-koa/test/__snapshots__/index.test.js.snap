// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`body handles errors 1`] = `
Array [
  Object {
    "message": "foo is required",
    "param": "foo",
    "required": true,
    "validator": "required",
  },
]
`;

exports[`query handles errors 1`] = `
Array [
  Object {
    "message": "foo is required",
    "param": "foo",
    "required": true,
    "validator": "required",
  },
]
`;
